### PR TITLE
fix(sequencers): broken invariant with unbonding sequencers

### PR DESCRIPTION
**Pull request by mtsitrin**
* fixes broken invariant once maxSequencer set to 0
* fixes check in sequencer creation, where `unbonding` sequencers doesn't count for `maxSequencer`

## Description

----

Closes [#714 ](https://github.com/dymensionxyz/dymension/issues/714)

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----;

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

File: x/sequencer/keeper/invariants.go [hunk 1]
```diff
@@ -76,7 +76,7 @@ func SequencersPerRollappInvariant(k Keeper) sdk.Invariant {
 			unbonding := k.GetSequencersByRollappByStatus(ctx, rollapp.RollappId, types.Unbonding)
 
 			// rollapp.MaxSequencers
-			if len(bonded)+len(unbonding) > int(rollapp.MaxSequencers) {
+			if rollapp.MaxSequencers > 0 && len(bonded)+len(unbonding) > int(rollapp.MaxSequencers) {
```
Comment (danwt) [inline x/sequencer/keeper/invariants.go:79]: can delete comment above

File: x/sequencer/keeper/msg_server_create_sequencer.go [hunk 2]
```diff
@@ -89,14 +89,15 @@ func (k msgServer) CreateSequencer(goCtx context.Context, msg *types.MsgCreateSe
...
-	currentNumOfSequencers := len(sequencersByRollapp)
-	if uint64(currentNumOfSequencers) >= rollapp.MaxSequencers && rollapp.MaxSequencers > 0 {
+	currentNumOfSequencers := len(bondedSequencers) + len(unbondingSequencers)
+	if rollapp.MaxSequencers > 0 && uint64(currentNumOfSequencers) >= rollapp.MaxSequencers {
 		return nil, types.ErrMaxSequencersLimit
 	}
 	// this is the first sequencer, make it a PROPOSER
-	if currentNumOfSequencers == 0 {
+	if len(bondedSequencers) == 0 {
 		sequencer.Proposer = true
```
Comment (danwt) [inline x/sequencer/keeper/msg_server_create_sequencer.go:101]: what about UNBONDED sequencers?
They will not be able to be created?

**Review by danwt — CHANGES_REQUESTED** [review]
I'm not sure how UNBONDED sequencers fit into this paradigm
Is there a source of truth definition for MaxSequencers?
Is it max active? max bonded? max total?

File: x/sequencer/keeper/msg_server_create_sequencer.go (hunk 2 above)
Comment (mtsitrin) [reply x/sequencer/keeper/msg_server_create_sequencer.go:101]: u can create only bonded sequencer. sequencer need to supply bond > min_bond

**Comment by mtsitrin** [comment]
> I'm not sure how UNBONDED sequencers fit into this paradigm Is there a source of truth definition for MaxSequencers? Is it max active? max bonded? max total?

unbonded sequencers not counted as active.
bonded are active
unbonding considered active until unbonded.

**Comment by danwt** [comment]
Ok, thanks for the explanation, can we document somewhere?

**Review by danwt — APPROVED** [review]
Approved
But I don't think it would do much harm to have a couple small keeper functions to return the `len`(s) and the total number, without needing to compute every time

**Review by zale144 — APPROVED** [review]

**Pull request merged**